examples/vision/super_resolution_sub_pixel.py

```diff
@@ -0,0 +1,513 @@
+"""
+
+"""
+### Prepare train and test data sequences.
+"""
+
+train_gen = BSDS500(
+    input_img_size=input_img_size,
+    target_img_size=target_img_size,
+    crop_size=crop_size,
+    input_img_paths=input_img_paths,
+    transform_input=transform_input,
+    transform_target=transform_target,
+    batch_size=batch_size,
+    upscale_factor=upscale_factor,
+    channels=channels,
+)
+
+valid_gen = BSDS500(
+    input_img_size=input_img_size,
+    target_img_size=target_img_size,
+    crop_size=crop_size,
+    input_img_paths=valid_img_paths,
+    transform_input=transform_input,
+    transform_target=transform_target,
+    batch_size=batch_size,
+    upscale_factor=upscale_factor,
+    channels=channels,
+)
+
+"""
+## Train the model with custom training loop.
+"""
+
+model = get_model(upscale_factor=upscale_factor, channels=channels)
+loss_fn = tf.keras.losses.MeanSquaredError()
+optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)
+
+import math
+import numpy as np
+
+epochs = 50
+
+best_psnr = 0
+for epoch in range(epochs):
+    print("\nStart of epoch %d" % (epoch + 1,))
+    total_psnr = 0
+    total_loss = 0
+    validation_loss = 0
+    # Iterate over the batches of the dataset.
+    for step, (x_batch_train, y_batch_train) in enumerate(train_gen):
+        with tf.GradientTape() as tape:
+            logits = model(x_batch_train, training=True)
+            loss_value = loss_fn(y_batch_train, logits)
+        grads = tape.gradient(loss_value, model.trainable_weights)
+        optimizer.apply_gradients(zip(grads, model.trainable_weights))
+        total_loss += loss_value
+        print(
+            "Training loss (for one batch) at step %d: %.4f" % (step, float(loss_value))
+        )
+    print(
+        "Avg. Loss for each epoch %d: %.4f" % (epoch + 1, total_loss / len(train_gen))
+    )
+    # Run a validation loop at the end of each epoch.
+    for x_batch_val, y_batch_val in valid_gen:
+        val_logits = model(x_batch_val, training=False)
+        val_loss_val = loss_fn(y_batch_val, val_logits)
+        # Calculate PSNR value.
+        psnr = 10 * math.log10(1 / val_loss_val)
+        total_psnr += psnr
+        validation_loss += val_loss_val
+    print("Avg. PSNR for each epoch: %.4f" % (total_psnr / len(valid_gen)))
+    print(
+        "Avg. Validation Loss for each epoch: %.4f" % (validation_loss / len(valid_gen))
+    )
+    best_psnr = max(best_psnr, float(total_psnr / len(valid_gen)))
+
+print("Best PSNR: %.4f" % best_psnr)
+
+"""
+Save model by `model.save()`
+"""
+
+model_path = os.path.join(os.getcwd(), "models")
+
+model.save(model_path)
+
+"""
+## Predict and plot the results.
+"""
+
+"""
+Import `mpl_tookit` to help us zoom in the image and compared with specific area.
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+import math
+
+from mpl_toolkits.axes_grid1.inset_locator import zoomed_inset_axes
+from mpl_toolkits.axes_grid1.inset_locator import mark_inset
+
+
+def plot_results(img, title):
```

Comment:
You could also put this logic in a callback, and save an image of the results to disk at the end of each epoch. It's great to see how the model progresses during training

Reply:
How do we get image input in callback if I use the `fit` ?  I didn't fin it on keras.io.

Reply:
Just pass a few image paths as constructor arguments to your callback.